File: loja-c/src/carrinho.c
#include "carrinho.h"

void carrinho_init(Carrinho *c) {
    c->quantidade = 0;
}

void carrinho_adicionar(Carrinho *c, Produto p) {
    if (c->quantidade < CARRINHO_MAX) {
        c->produtos[c->quantidade++] = p;
    }
}

double carrinho_total(const Carrinho *c) {
    double total = 0.0;

    for (int i = 0; i < c->quantidade; i++) {
        total += c->produtos[i].preco;
    }

    return total;
}

File: loja-rust/src/carrinho.rs
use crate::produto::Produto;

#[derive(Debug, Default)]
pub struct Carrinho {
    produtos: Vec<Produto>,
}

impl Carrinho {
    pub fn new() -> Self {
        Self {
            produtos: Vec::new(),
        }
    }

    pub fn adicionar(&mut self, produto: Produto) {
        self.produtos.push(produto);
    }

    pub fn total(&self) -> f64 {
        self.produtos.iter().map(|p| p.preco).sum()
    }

    pub fn quantidade(&self) -> usize {
        self.produtos.len()
    }
}
